WalletWasabi.Tests/UnitTests/Crypto/ProofSystemTests.cs

```diff
@@ -0,0 +1,51 @@
+using System;
+using NBitcoin.Secp256k1;
+using WalletWasabi.Crypto;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.Crypto
+{
+	public class ProofSystemTests
+	{
+		[Fact]
+		[Trait("UnitTest", "UnitTest")]
+		public void CanProveAndVerifyMAC()
+		{
+			// The coordinator generates a composed private key called CoordinatorSecretKey 
+			// and derives from that the coordinator's public parameters called CoordinatorParameters.
+			var rnd = new SecureRandom();
+			var coordinatorKey = new CoordinatorSecretKey(rnd);
+			var coordinatorParameters = coordinatorKey.ComputeCoordinatorParameters();
+
+			// A blinded amout is known as an `attribute`. In this case the attribute Ma is the 
+			// valued 10000 blinded with a random `blindingFactor`. This attribute is sent to 
+			// the coordinator.
+			var amount = new Scalar(10_000);
+			var blindingFactor = rnd.GetScalar();
+			var Ma = amount * Generators.G + blindingFactor * Generators.Gh;
+
+			// The coordinator generates a MAC and a proof that that MAC was generated using the 
+			// coordinator's secret key. The coordinator sends the pair (MAC + proofOfMac) back 
+			// to the client.
+			var t = rnd.GetScalar();
+			var mac = MAC.ComputeMAC(coordinatorKey, Ma, t);
+
+			var coordinatorStatement = ProofSystem.CreateStatement(coordinatorParameters.Cw, coordinatorParameters.I, mac.V, MAC.GenerateU(t), Ma, t);
+			var proverBuilder = ProofSystem.CreateProver(coordinatorStatement.ToKnowledge(coordinatorKey.GetWitness(t)));
```

Comment:
My understanding is that `coordinatorKey.GetWitness(t)` returns `{w, wp, x0, x1, ya}`. I think it is confusing that GetWitness() has an argument, which is public knowledge anyway. If it is just a simple getter function, then it should not have an argument, right?

Reply:
it returns `t` as well, but that is a mistake, `t` should not be a part of the witness and the 5th generator can be removed from each equation (see https://github.com/zkSNACKs/WalletWasabi/pull/4350/files#r486638509)

but even after `t` is removed as an argument i think `GetWitness` has a problem - it spreads knowledge about the order of the terms between two unrelated files, `ProofSystem` knows about the order because it sets the order of the corresponding generators, and `CoordinatorSecretKey` knows about it because it converts from named properties to a `ScalarVector`. I think that both should be handled in `ProofSystem`.

Reply:
I have changed that.